Validation pipeline with testing

EVWorth opened the pull request:
closes #1204 

bumped checkout version, added steps to install uv, setup the python env and execute all pytests

@cayossarian I was going to figure out how to mock a failure to demonstrate but it does seem like one of the tests is actually failing (my branch is up to date with main) https://github.com/EVWorth/alarmo/actions/runs/16078828288/job/45379758206?pr=3

I'm not super familiar with pytest so I'm not sure if there's a way to tune the logs returned.. seems like it throws quite a bit of data when something goes wrong, but I guess that's a good thing.

This might also make sense as a separate tests workflow, so that tests are allowed to hard fail separately than lets say ruff formatting validation.

thoughts? 

ps.. also added workflow_dispatch to allow for manual runs of the workflow from actions

EVWorth commented:
@cayossarian I attempted to do some testing and am at a bit of a loss. 

I tried 2 things. first, I tried uv sync --upgrade to bump all package versions, which didnt resolve the test error.

I then tried force tagging the versions that were documented in the pyproject toml to be ==...
![image](https://github.com/user-attachments/assets/d641b94f-62be-432a-b170-4374ced8ab03)

and attempted to run the tests again.. test_last_triggered_attribute_set_on_alarm_trigger continued to fail.

taking a look at this test, I threw it into debug mode. 

![image](https://github.com/user-attachments/assets/db5740b7-f4b7-42a9-97c6-6b870a11cc80)


as you can see, datetime.now() returned 2025/07/04 22:50:40 (10:50pm)
but the last_triggered time is set to 20:50 (8:50pm)

it almost seems like the last triggered time is being evaluated in PST instead of my CST.

Not sure what the issue is as I got lost trying to dig into where last_triggered is being set from.

this behavior was consistent across all versions of the packages I tried installing.

cayossarian commented:
What I did right off was remove my local repo entirely, sync's my GitHub fork with Neil's, re-cloned my local copy, and everything tested fine.   I could have just removed the venv but I wanted to test whether Neil's repo was solid.  When I reviewed your last PR I probably just ran the tests command without updating the dependencies with that errant flag.  

If you update dependencies with --update all your local virtual environment dependencies have the latest packages which HA won't work with.  If you didn't have a a virtual environment and you used the system environment then that now has all the latest packages.  

Once you update the packages to the latest version you must get rid of them either remove the virtual environment or downgrade them by force.

EVWorth commented:
So... I tried a few things:

note: before all testing I confirmed my forked main branch was up to date with upstream main.
all testing was done on my synced fork main branch. 

1) Played around with deleting the venv and running uv venv and uv sync.
 - test failed
2) deleted venv, went and found the old requirements-test.txt file and uv pip installed all requirements. 
-  test failed
3) deleted my local alarmo repo, git cloned https://github.com/nielsfaber/alarmo.git, ran uv venv, (confirmed no packages were installed with uv pip freeze) ran uv sync. confirmed uv.lock file was not updated and that packages installed matched what is in the uv.lock from nielsfaber/alarmo.git@main. 
- test failed. 

can you walk me through your setup steps so I can try to reproduce? I feel like I did the same thing but it's still failing.


also...

> If you update dependencies with --update all your local virtual environment dependencies have the latest packages which HA won't work with. If you didn't have a virtual environment and you used the system environment, then that now has all the latest packages.

this shouldn't be an issue. UV evaluates package dependencies at --update time. what's in the uv.lock file should be compatible with the lowest common denominator for package dependencies and would throw an error if one package needed a higher minimum version than another's max version. 
Where you able to pinpoint which package was causing the issue? all I saw with that test were what looked like weird timezone offsets. I even tried pinning the required versions of the software to what was minimally required in the requirements-test.txt file and the test still failed.

I can be a little slow sometimes, so happy to be corrected :), but it seems to me the test is failing regardless of what package versions are installed.

EVWorth commented:
ok now im really at a loss. I've been using WSL 2.0 for all my code so far.

just to make sure it wasn't something weird going on with my setup, I did the following:
1) downloaded virtual box
2) installed Ubuntu Server 24.04.02 LTS
3) sudo apt install pipx
4) python3 -m pipx uv
5) cloned https://github.com/cayossarian/alarmo.git
6) git checkout origin/correct_test_readme
7) uv venv
8) uv sync
9) uv run pytest tests

![image](https://github.com/user-attachments/assets/9dc42631-ef10-4361-b553-6c4ecb4d23e0)

😭

cayossarian commented:
The error is actually not a test failure per se it's  a lingering time which means the test did quit with a timer still active but it should not happen as the tests have clean up routines.  But yeah odd.

which version of python is the venv using?  did you source the venv first before using uv sync, etc so the packages get installed there and the subsequent uv calls get those correct packages?

HA and the custom component/pytest are using ❯ python --version
Python 3.13.2
which python should resolve to the venv before you run uv

cayossarian commented:
Yes you're correct.  Let me clean up that timer in my PR and try in a sandbox.

cayossarian commented:
I changed my PR to draft and simplified the test in question, while leaving the timer cleanup, pushing those changes up to my PR branch.  Also updated the dependencies for some of the packages and HA.  To save some time for Neil can you pull the PR again and that way if you can run the tests successfully  you can merge my PR and I can close mine.  I'll then try your pipeline in a sandbox.  Thanks

EVWorth commented:
@cayossarian I pulled down your correct_test_readme branch and ran uv run pytest tests and all completed successfully :)

note: in tests/README.md id recommend removing:

> ### **Note:** ensure you commit any changes to the uv.lock file to ensure the main branch is updated to use the newest version of all packages

since you removed the --upgrade  from uv sync

cayossarian commented:
Yes, the readme has `uv sync` so upgrade is not desirable.

Ideally, it would be helpful to have pre-commit checks run locally before a commit succeeds.  When files are staged and user tries to commit locally the tests will run so you don't have to wait until the workflow fails on GitHub (which is what happened when you uploaded this last PR and the test failed.  That's how HA core works.  I have pre-commit on all the [SpanPanel](https://github.com/SpanPanel) repositories so if you're not familiar with how to do that from your branch.  You can then merge those changes into this PR so we don't need another.

On ruff I think Neil needs to be on board since the code would need modifications to be PEP8 compliant or we would need to relax the constraints so much there would be no point.  The tests would provide assurance but that's Neils call, he may be happy to leave it as is.

EVWorth commented:
I'm down for that! I'll work on adding pre-commit and a pytest step

Do you think there's still value in having a pytest validation workflow? Technically pre-commit only runs if you set it up, so there'd be nothing stopping someone from submitting a pr without precommit validation. 

I agree on ruff. My stance would be standard implementation with full pep8 compliance across the board. 

It'd mean basically every Python script gets a complete overhaul formatting wise but then we can build in precommit checks and validation workflows

EVWorth commented:
@cayossarian I may be missing some things still, but:

added pre-commit to the pyproject toml, and set up a setup.sh file.. primarily to get git/vscode setup to use uv run pre-commit whenever a commit is called.

this was the only way I could figure out how to set it up so that pre-commit would ~just work~ only other option (I think?) would be to have devs install pre-commit on system and then run pre-commit install

happy for feedback or alterations. I've given you access to my fork in case you'd like to collaborate on this PR!

EVWorth commented:
@cayossarian I was overcomplicating things a bit. no need for a setup script. 

I updated the documentation:
![image](https://github.com/user-attachments/assets/19bfa491-ce40-4da0-8f20-b2ba7b327f2e)


note.. I also removed the --upgrade from the uv sync command in the tests.md file and deleted the relevant test code to mirror what you'd done in your branch, so that I could confirm that pre-commit could complete successfully.

I saw you closed your PR for that branch, I assume since I'd ended up essentially merging your changes in my branch from testing? or do I need to restore the code I deleted from your test?

cayossarian commented:
@nielsfaber At this point I think it's best we align on how you want your pre-commits and/or pipeline to work.  There are a couple of options and we can provide this effort in this PR or you can do it on your own.

- Force pre-commits to run on staged during commit attempt and fail the commit if the tests don't pass (the HA core model)
- Run the CI on GitHub workflow where the tests must pass
- Do both (the HA model)
- Run the tests locally on-demand

How would you like to proceed?

EVWorth commented:
> @EVWorth you could squash a bunch of those commits where you reverted some things in and out to the make the commit history cleaner and force push to the branch.

squashed! thanks, I've never done that before. interesting to learn how!

**note:** 
> This PR has the pre-commits that can be run manually and fixed a test that had a lingering timer. The PR could be merged as is but it doesn't enforce passing tests or any other checks prior to commit.


if I update a test to fail an assertion:
![image](https://github.com/user-attachments/assets/68ef1422-5003-42aa-8e4c-618cd9e54397)
and try to commit it, it does enforce passing the tests prior to committing.

from vscode:
![image](https://github.com/user-attachments/assets/3a04abd1-2908-4d55-9cf9-ab25951e1c54)
![image](https://github.com/user-attachments/assets/4e209475-23a7-4141-bfd6-2cfdfcef56bb)

from commandline:
![image](https://github.com/user-attachments/assets/1d1ac5bd-409b-4f08-b704-1bc33777a910)
![image](https://github.com/user-attachments/assets/b85e472c-695c-4b9e-bbfa-0b756eb64520)

cayossarian commented:
Try changing and committing something in the Alarmo that would fail a test and see if it runs the tests prior to commit and fails to commit, putting the errant file back in staged.

EVWorth commented:
example in init.py:
changing sensor_handler to sensor_handlers
![image](https://github.com/user-attachments/assets/28db01b2-e119-4dcc-aefd-e2c3e9c422a0)

begin commit, pytests running
![image](https://github.com/user-attachments/assets/e214941b-bdd0-45d8-811b-9e5475f52825)

pytest fails, init.py still staged (not committed)

![image](https://github.com/user-attachments/assets/a30a7f1c-8eb0-496c-b6d0-cf3bfa903c98)

cayossarian commented:
Perfect!

EVWorth commented:
This PR should be ready to go. 

@nielsfaber all tests will be run by the developer every time they commit code to their branch, and the tests will be re-run inside the PR when it is first opened or new commits are added to the branch in origin. 

<img width="988" height="193" alt="image" src="https://github.com/user-attachments/assets/2a7ed62e-2f5a-4e8c-ab9f-936c029bb936" />

<img width="919" height="1249" alt="image" src="https://github.com/user-attachments/assets/70e886a6-c1d9-48e6-afb4-e7b853188147" />

if the tests fail in the PR, you'll see a failed check and can review the output, but this shouldn't happen if people are using pre-commit :) 

I think this one is ready to be reviewed/merged!!

nielsfaber commented:
@EVWorth This looks very promising, I'm very curious to see it in action!
Thanks a million for for contributing this, it's a very valuable addition to the tests contributed by @cayossarian, a big improvement to the quality of this project.
I made an attempt reviewing the whole PR, but to me all looks good. If further refinement is needed I hope I can get in touch with you again.

EVWorth commented:
For sure! Just @ me if you need something!

nielsfaber commented:
@EVWorth As of this week the github validation action is failing due to dependency of the 'pillow' package, of which the installation appears to be failing. 
I have attached the log with the error description, otherwise this can also be [found here](https://github.com/nielsfaber/alarmo/actions/runs/18569010061/job/52937794175).
Do you have any idea what causes the problem? 

[github_action_log.txt](https://github.com/user-attachments/files/22955467/github_action_log.txt)
